drivers: regulator: refine API

gmarull opened the pull request:
This PR refines the regulator API in multiple aspects: available operations, interface design, docs, etc.

Please check each commit for more details.

gmarull committed:
drivers: regulator: improve regulator_count_voltages specs

- Interface guarantees that voltage indices start from zero to
  regulator_count_voltages()
- Function always return an unsigned number. If the device doesn't
  implement the API returned value is zero, as expected.

Signed-off-by: Gerard Marull-Paretas <gerard.marull@nordicsemi.no>

gmarull committed:
drivers: regulator: drop regulator_count_modes

The regulator_count_modes was an API that was not useful. Unlike with
voltages where the interface guarantees they are zero-indexed, modes can
take an arbitrary identifier. So counting supported modes doesn't
provide any useful information such as if a mode is allowed.

Signed-off-by: Gerard Marull-Paretas <gerard.marull@nordicsemi.no>

gmarull committed:
drivers: regulator: improve regulator_list_voltages

- Rename to regulator_list_voltage (it is listing a single voltage)
- Function returns the value via a parameters, so that we can indicate
  wether the given index is valid or not. If a driver doesn't implement
  this call, function returns -EINVAL (as it should be).

Signed-off-by: Gerard Marull-Paretas <gerard.marull@nordicsemi.no>

gmarull committed:
drivers: regulator: provide generic regulator_is_supported_voltage

The function can be implemented by using regulator_count_voltages() +
regulator_list_voltage(), so there's no need to defer the job to each
driver.

Signed-off-by: Gerard Marull-Paretas <gerard.marull@nordicsemi.no>

gmarull committed:
drivers: regulator: clarify regulator_set_voltage

- Voltage is applied to the active or selected mode
- Implementation must return -EINVAL if given voltage window is not
  valid.

Signed-off-by: Gerard Marull-Paretas <gerard.marull@nordicsemi.no>

gmarull committed:
drivers: regulator: drop mode specific APIs

- Mode specific APIs repeat the same functionality offered by non-mode
  specific APIs
- The same functionality can be achieved by the non-mode APIs, since
  they apply to the active mode which can be set using
  regulator_set_mode() first.

Signed-off-by: Gerard Marull-Paretas <gerard.marull@nordicsemi.no>

gmarull committed:
drivers: regulator: improve regulator_get_voltage

- Function now returns error, value is obtained by reference. This
  allows to propagate potential bus errors.

Signed-off-by: Gerard Marull-Paretas <gerard.marull@nordicsemi.no>

gmarull committed:
drivers: regulator: clarify regulator_set_mode interface

Clarify the API for regulators that have the option to set mode
externally, such as PCA9420. Adjust the PCA9420 driver to comply with
the interface.

Signed-off-by: Gerard Marull-Paretas <gerard.marull@nordicsemi.no>

gmarull committed:
drivers: regulator: improve regulator_get_current_limit

- Function returns now the value by reference, similar to voltage
  counterparts.

Signed-off-by: Gerard Marull-Paretas <gerard.marull@nordicsemi.no>

gmarull committed:
drivers: regulator: clarify regulator_set_current_limit()

- Clarify the documentation, add -EINVAL return value when current limit
  is out of the window.

Signed-off-by: Gerard Marull-Paretas <gerard.marull@nordicsemi.no>

gmarull committed:
drivers: regulator: add regulator_mode_t opaque type

Define regulator mode as an opaque type, same as we do in other
subsystems like GPIO (e.g. gpio_flags_t).

Signed-off-by: Gerard Marull-Paretas <gerard.marull@nordicsemi.no>

gmarull committed:
drivers: regulator: pca9420: some minor improvements

- Remove redundant comments
- Remove internal *_mode functions, as they were only used once.

Signed-off-by: Gerard Marull-Paretas <gerard.marull@nordicsemi.no>

danieldegrasse commented on drivers/regulator/regulator_pca9420.c:
```diff
@@ -322,76 +322,6 @@ static int regulator_pca9420_get_current_limit(const struct device *dev)
 	return MIN(config->max_ua, cconfig->vin_ilim_ua);
 }
 
-/*
```
I don't think it makes sense to drop these mode specific configuration APIs. For example, the PCA9420 boots with different target voltages for each mode. So consider the following (toy) example:

- User designs system where SW1 output (defaults to 1.0V at powerup) is connected to a device that runs in the range 1.0-1.1V.
- User wants to utilize multiple modes in their application, and selects mode 1 as their second target mode
- Since SW1 defaults to an output voltage of 1.2V, the user effectively cannot switch to this mode.

I hope this makes it clear why I added mode support using these APIs initially. While this isn't a likely example, I still think the ability to configure target voltages before applying them is valuable to the user.

gmarull replied:
Modes are defined as different operating modes by the regulator. For example, in case of nPM6001 you have hysteretic or PWM mode. Hysteretic is efficient at low load currents, whereas PWM is efficient at higher current loads.

![image](https://user-images.githubusercontent.com/25011557/205872098-02fe5cb4-6333-4d4c-a4bf-19d439e641a9.png)

So for example, in low-power mode you'll likely consume a few u/mA so hysteretic will give you the maximum efficiency and so battery life. When active, PWM mode will likely be the best choice. Linux even has an [API](https://elixir.bootlin.com/linux/latest/source/drivers/regulator/core.c#L4644) where you can specify your current load, so that the driver can choose the optimum mode internally.

From Linux regulator framework slides here: https://elinux.org/images/c/c1/Regulator-api-celf.pdf
![image](https://user-images.githubusercontent.com/25011557/205871327-b230120c-189e-42b6-b4de-b1cf6548eeae.png)

In case of PCA9420 I haven't seen any explanation of what modes 0...3 really mean. If they are just preprogrammed states, we may need to adjust the API for such functionality.

For reference, other examples:
https://www.ti.com/lit/ds/symlink/tps62866.pdf
https://elixir.bootlin.com/linux/latest/source/include/dt-bindings/regulator/ti,tps62864.h

danieldegrasse replied:
Thank you for the clarification here. I was under the impression that modes were generally understood to be very hardware specific, and have different meanings across different regulators. If we want to use modes for operating efficiency at different loads, than the PCA9420 does not have modes in that sense.

The primary use case I am aware of for the modes present on the PCA9420 is automatically transitioning between groups of set voltages when changing power modes. On the RT685 and RT595, there are two output pins on the SOC that can be used to indicate the power state. The idea is that when paired with the PCA9420, these RT MCUs can seamlessly transition between power states, provided the voltages have been configured for each target mode:
![image](https://user-images.githubusercontent.com/34665051/205960982-54b46c1d-c82d-4958-9c45-532b87a52fa4.png)

If we want to use another API to configure these modes on the PCA9420, I'm open to that. Really, my primary concern is how to effectively enable this use case, which is required to optimize power consumption on these MCUs.

For more info on the power management optimization for RT595: https://www.nxp.com/docs/en/application-note/AN13162.pdf

gmarull replied:
Thanks for all these details, I'll try to introduce some changes to the API to cover this case while keeping the usage of modes.

danieldegrasse replied:
Hi @gmarull, I've taken a look. I'd prefer that some standard API exists to cover the PCA9420's use case. While this is not a common feature, other regulators do exist that implement voltage scaling based on GPIO pin settings. Could we consider using `state` to describe the APIs present here? So these regulator APIs would be renamed to `regulator_set_state_voltage`, `regulator_set_state`, and `regulator_state_disable/enable`

gmarull replied:
> Hi @gmarull, I've taken a look. I'd prefer that some standard API exists to cover the PCA9420's use case. While this is not a common feature, other regulators do exist that implement voltage scaling based on GPIO pin settings. Could we consider using `state` to describe the APIs present here? So these regulator APIs would be renamed to `regulator_set_state_voltage`, `regulator_set_state`, and `regulator_state_disable/enable`

OK, let me see what I can do. Note that the state in case of PCA9420 applies to all regulators, so I feel we need a separate API for the "parent regulator". Could you list other PMICs with this feature so I make sure what I propose solves more than a single usecase?

danieldegrasse commented on drivers/regulator/regulator_pca9420.c:
```diff
@@ -271,23 +271,14 @@ static unsigned int regulator_pca9420_count_voltages(const struct device *dev)
- */
-static int32_t regulator_pca9420_list_voltages(const struct device *dev,
-					       unsigned int selector)
+static int regulator_pca9420_list_voltage(const struct device *dev,
```
Just a note here: this API (and several APIs changed in later commits) are based on the Linux regulator API: https://docs.kernel.org/driver-api/regulator.html#c.regulator_list_voltage. I'm aware Zephyr isn't Linux, but I wanted to highlight the original reasoning for the design here. I'm not opposed to the API change though.

gmarull replied:
One option could be to return the value, but then all regulators need to implement the capability of listing voltages, even simple GPIO switches. It may actually make sense to specify the voltage on switches as well, after all they turn on/off a power rail that has a fixed voltage.

Let me think about this and come up with a proposal.

gmarull replied:
Considering that the voltage domain is (-INT32_MAX, INT32_MAX), we have no other choice but to return the value by reference. Otherwise, we have no way to indicate whether the given index is valid or not. On Linux, only positive voltages are supported by the API, in our case both positive/negative are supported.

danieldegrasse reviewed (changes requested):
In general not opposed to the API changes here, but I have concerns about dropping the set_mode API (see below)

gmarull committed:
drivers: regulator: fix/improve usage of devicetree properties

Most of devicetree properties for regulator, such as:

- regulator-min/max-microvolt
- regulator-min/max-microamp
- regulator-allowed-modes
- etc.

Are meant to specify limits on what consumers may set. They are **NOT**
meant to describe the hardware capabilities. For example, I could have a
BUCK converter that supports 0-5V output voltage, but my circuit may
only allow working on the 2.7-3.3V range.

This patch reworks the API so that the API class layer manages this
information. This is done by drivers collecting all such fields in a
common configuration structure that is later accessed by the class
layer. This simplifies drivers implementation. For example, if A
consumer calls regulator_set_voltage() with a voltage that is supported
but not allowed, driver code won't be called. Similarly, if a regulator
is configured to be `always-on`, enable/disable driver code will never
be called.

Drivers have been adjusted. PCA9420 mode settings have been removed from
devicetree in this commit as they are not actual modes but PMIC states.
This will be refactored in a follow-up commit.

Signed-off-by: Gerard Marull-Paretas <gerard.marull@nordicsemi.no>

gmarull committed:
drivers: regulator: add parent DVS API

PCA9420 PMIC offers of multiple operation states, or DVS (Dynamic
Voltage Scaling). Such states may be automatically changed by hardware
using MODESEL0/1 pins. Certain MCUs allow to automatically configure
certain output pins when entering low power modes so that PMIC state is
changed without software intervention.  This means that application just
needs to configure the voltages for each state using
`nxp,modeN-microvolt`, set `nxp,enable-modesel-pins` in devicetree and
forget about configuring regulators.

This patch introduces a new _parent_ API to expose such functionality in
a vendor agnostic way. Consider this API as experimental for now, until
we have other usecases.

Signed-off-by: Gerard Marull-Paretas <gerard.marull@nordicsemi.no>

gmarull committed:
drivers: regulator: add regulator_is_enabled

Add a new API to check if a regulator is enabled or not.

Signed-off-by: Gerard Marull-Paretas <gerard.marull@nordicsemi.no>

gmarull committed:
drivers: regulator: add common init enable API

Add a new API for drivers that can be called to initialize the regulator
at init time if `regulator-boot-on` or `regulator-always-on` are set.

Signed-off-by: Gerard Marull-Paretas <gerard.marull@nordicsemi.no>

gmarull committed:
drivers: regulator: add API to read error flags

Add a new API to obtain the active error flags on a regulator, e.g.
overcurrent, overvoltage or over temperature.

Signed-off-by: Gerard Marull-Paretas <gerard.marull@nordicsemi.no>

danieldegrasse reviewed (approved):
Changes look good to me, thank you for the updates
